src/runtime/FinalizationRegistryObject.cpp

```diff
@@ -0,0 +1,91 @@
+
+#include "Escargot.h"
+#include "FinalizationRegistryObject.h"
+#include "Context.h"
+
+namespace Escargot {
+
+FinalizationRegistryObject::FinalizationRegistryObject(ExecutionState& state, EncodedValue cleanupCallback, Object* realm)
+    : FinalizationRegistryObject(state, state.context()->globalObject()->finalizationRegistryPrototype(), cleanupCallback, realm)
+{
+}
+
+FinalizationRegistryObject::FinalizationRegistryObject(ExecutionState& state, Object* proto, EncodedValue cleanupCallback, Object* realm)
+    : Object(state, proto)
+    , m_cleanupCallback(cleanupCallback)
+    , m_realm(realm)
+{
+}
+
+void* FinalizationRegistryObject::FinalizationRegistryObjectItem::operator new(size_t size)
+{
+    static bool typeInited = false;
+    static GC_descr descr;
+    if (!typeInited) {
+        GC_word obj_bitmap[GC_BITMAP_SIZE(FinalizationRegistryObject::FinalizationRegistryObjectItem)] = { 0 };
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, weakRefTarget));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, heldValue));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, unregisterToken));
+        descr = GC_make_descriptor(obj_bitmap, GC_WORD_LEN(FinalizationRegistryObject::FinalizationRegistryObjectItem));
+        typeInited = true;
+    }
+    return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
+}
+
+void* FinalizationRegistryObject::operator new(size_t size)
+{
+    static bool typeInited = false;
+    static GC_descr descr;
+    if (!typeInited) {
+        GC_word obj_bitmap[GC_BITMAP_SIZE(FinalizationRegistryObject)] = { 0 };
+        Object::fillGCDescriptor(obj_bitmap);
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject, m_cleanupCallback));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject, m_realm));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject, m_cells));
+        descr = GC_make_descriptor(obj_bitmap, GC_WORD_LEN(FinalizationRegistryObject));
+        typeInited = true;
+    }
+    return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
+}
+
+void FinalizationRegistryObject::setCell(ExecutionState& state, const Value& weakRefTarget, const Value& heldValue, const Value& unregisterToken)
+{
+    auto newCell = new FinalizationRegistryObjectItem();
+    newCell->weakRefTarget = weakRefTarget;
+    newCell->heldValue = heldValue;
+    newCell->unregisterToken = unregisterToken;
+    m_cells.pushBack(newCell);
+}
+
+bool FinalizationRegistryObject::deleteCell(ExecutionState& state, Value unregisterToken)
+{
+    bool removed = false;
+    for (size_t i = 0; i < m_cells.size(); i++) {
+        if (!m_cells[i]->unregisterToken.isEmpty() && unregisterToken.equalsToByTheSameValueAlgorithm(state, m_cells[i]->unregisterToken)) {
+            m_cells.erase(i);
+            i--;
+            removed = true;
```

Comment:
Is it still necessary to check empty value (`!m_cells[i]->unregisterToken.isEmpty()`) ?
And do we need to iterate the entire cells even after we find one correspond element and erase it?
I mean that is there any possible case that same elements are inserted in the cells?

Reply:
Sadly the standard says the following:
` 6. For each Record { [[Target]], [[Holdings]], [[UnregisterToken]] } cell that is an element of finalizationRegistry.[[Cells]], do
    a. If SameValue(cell.[[UnregisterToken]], unregisterToken) is true, then
      i. Remove cell from finalizationRegistry.[[Cells]].
      ii. Set removed to true.`
      
 If we return after a successfully finding a member and removing a few test-cases will fail. 
For Example: 
  built-ins/FinalizationRegistry/prototype/unregister/custom-this.js
  built-ins/FinalizationRegistry/prototype/unregister/unregister.js

Reply:
Thanks for your detailed explanation.